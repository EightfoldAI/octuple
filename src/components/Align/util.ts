import ResizeObserver from 'resize-observer-polyfill';
import { canUseDocElement, contains } from '../../shared/utilities';
import type { TargetPoint } from './Align.types';

export function isSamePoint(prev: TargetPoint, next: TargetPoint) {
  if (prev === next) return true;
  if (!prev || !next) return false;

  if ('pageX' in next && 'pageY' in next) {
    return prev.pageX === next.pageX && prev.pageY === next.pageY;
  }

  if ('clientX' in next && 'clientY' in next) {
    return prev.clientX === next.clientX && prev.clientY === next.clientY;
  }

  return false;
}

export function restoreFocus(activeElement: any, container: any) {
  // Focus back if is in the container
  if (
    canUseDocElement() &&
    activeElement !== document.activeElement &&
    contains(container, activeElement) &&
    typeof activeElement.focus === 'function'
  ) {
    activeElement.focus();
  }
}

export function onViewportResize(element: HTMLElement, callback: Function) {
  let prevWidth: number = null;
  let prevHeight: number = null;

  function onResize([{ target }]: ResizeObserverEntry[]) {
    if (!canUseDocElement() || !document.documentElement.contains(target)) {
      return;
    }
    const { width, height } = target.getBoundingClientRect();
    const fixedWidth = Math.floor(width);
    const fixedHeight = Math.floor(height);

    if (prevWidth !== fixedWidth || prevHeight !== fixedHeight) {
      // https://webkit.org/blog/9997/resizeobserver-in-webkit/
      Promise.resolve().then(() => {
        callback({ width: fixedWidth, height: fixedHeight });
      });
    }

    prevWidth = fixedWidth;
    prevHeight = fixedHeight;
  }

  const resizeObserver: ResizeObserver = new ResizeObserver(onResize);
  if (element) {
    resizeObserver.observe(element);
  }

  return () => {
    resizeObserver?.disconnect();
  };
}
