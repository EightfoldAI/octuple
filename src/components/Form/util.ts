import type { InternalOcNamePath } from './Internal/OcForm.types';

// Form item name black list. In form, you can use form.id get the form item element.
// use object hasOwnProperty will get better performance if black list is longer.
const formItemNameBlackList = ['parentNode'];

// default form item id prefix.
const defaultItemNamePrefix: string = 'form_item';

export function toArray<T>(candidate?: T | T[] | false): T[] {
    if (candidate === undefined || candidate === false) return [];

    return Array.isArray(candidate) ? candidate : [candidate];
}

export function getFieldId(
    namePath: InternalOcNamePath,
    formName?: string
): string | undefined {
    if (!namePath.length) return undefined;

    const mergedId = namePath.join('_');

    if (formName) {
        return `${formName}_${mergedId}`;
    }

    const isIllegalName = formItemNameBlackList.indexOf(mergedId) >= 0;

    return isIllegalName ? `${defaultItemNamePrefix}_${mergedId}` : mergedId;
}
