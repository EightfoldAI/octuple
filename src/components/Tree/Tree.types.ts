import type {
  BasicDataNode,
  DataNode,
  Key,
  OcTreeProps,
} from './Internal/OcTree.types';

export type SwitcherIcon =
  | React.ReactNode
  | ((props: { expanded: boolean }) => React.ReactNode);

export interface OcTreeNodeAttribute {
  eventKey: string;
  classNames: string;
  expanded: boolean;
  selected: boolean;
  checked: boolean;
  halfChecked: boolean;
  children: React.ReactNode;
  title: React.ReactNode;
  pos: string;
  dragOver: boolean;
  dragOverGapTop: boolean;
  dragOverGapBottom: boolean;
  isLeaf: boolean;
  selectable: boolean;
  disabled: boolean;
  disableCheckbox: boolean;
}

export interface OcTreeNodeProps {
  classNames?: string;
  checkable?: boolean;
  disabled?: boolean;
  disableCheckbox?: boolean;
  title?: string | React.ReactNode;
  key?: Key;
  eventKey?: string;
  isLeaf?: boolean;
  checked?: boolean;
  expanded?: boolean;
  loading?: boolean;
  selected?: boolean;
  selectable?: boolean;
  icon?: ((treeNode: OcTreeNodeAttribute) => React.ReactNode) | React.ReactNode;
  children?: React.ReactNode;
  [customProp: string]: any;
}

export interface OcTreeNode extends React.Component<OcTreeNodeProps, {}> {}

export interface OcTreeNodeBaseEvent {
  node: OcTreeNode;
  nativeEvent: MouseEvent;
}

export interface OcTreeNodeCheckedEvent extends OcTreeNodeBaseEvent {
  event: 'check';
  checked?: boolean;
  checkedNodes?: OcTreeNode[];
}

export interface OcTreeNodeSelectedEvent extends OcTreeNodeBaseEvent {
  event: 'select';
  selected?: boolean;
  selectedNodes?: DataNode[];
}

export interface OcTreeNodeExpandedEvent extends OcTreeNodeBaseEvent {
  expanded?: boolean;
}

export interface OcTreeNodeMouseEvent {
  node: OcTreeNode;
  event: React.DragEvent<HTMLElement>;
}

export interface OcTreeNodeDragEnterEvent extends OcTreeNodeMouseEvent {
  expandedKeys: Key[];
}

export interface OcTreeNodeDropEvent {
  node: OcTreeNode;
  dragNode: OcTreeNode;
  dragNodesKeys: Key[];
  dropPosition: number;
  dropToGap?: boolean;
  event: React.MouseEvent<HTMLElement>;
}

// [Legacy] Compatible for v3
export type TreeNodeNormal = DataNode;

export type DraggableFn = (node: OcTreeNode) => boolean;

export interface DraggableConfig {
  icon?: React.ReactNode | false;
  nodeDraggable?: DraggableFn;
}

export interface TreeProps<T extends BasicDataNode = DataNode>
  extends Omit<OcTreeProps<T>, 'showLine' | 'direction' | 'draggable'> {
  showLine?: boolean | { showLeafIcon: boolean };
  classNames?: string;
  multiple?: boolean;
  autoExpandParent?: boolean;
  checkStrictly?: boolean;
  checkable?: boolean;
  disabled?: boolean;
  defaultExpandAll?: boolean;
  defaultExpandParent?: boolean;
  defaultExpandedKeys?: Key[];
  expandedKeys?: Key[];
  checkedKeys?: Key[] | { checked: Key[]; halfChecked: Key[] };
  defaultCheckedKeys?: Key[];
  selectedKeys?: Key[];
  defaultSelectedKeys?: Key[];
  selectable?: boolean;
  filterOcTreeNode?: (node: OcTreeNode) => boolean;
  loadedKeys?: Key[];
  draggable?: DraggableFn | boolean | DraggableConfig;
  style?: React.CSSProperties;
  showIcon?: boolean;
  icon?:
    | ((nodeProps: OcTreeNodeAttribute) => React.ReactNode)
    | React.ReactNode;
  switcherIcon?: SwitcherIcon;
  children?: React.ReactNode;
  blockNode?: boolean;
  virtual?: boolean;
}
