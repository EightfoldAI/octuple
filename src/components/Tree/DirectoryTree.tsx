import React from 'react';
import { debounce, mergeClasses } from '../../shared/utilities';
import type OcTree from './Internal';
import { conductExpandParent } from './Internal/util';
import type { EventDataNode, DataNode, Key } from './Internal/OcTree.types';
import {
  convertDataToEntities,
  convertTreeToData,
} from './Internal/utils/treeUtil';
import type { TreeProps, OcTreeNodeAttribute } from './Tree.types';
import Tree from './Tree';
import { calcRangeKeys, convertDirectoryKeysToNodes } from './Utils/dictUtil';
import { Icon, IconName, IconSize } from '../Icon';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';

import styles from './Styles/tree.module.scss';

export type ExpandAction = false | 'click' | 'doubleClick';

export interface DirectoryTreeProps extends TreeProps {
  expandAction?: ExpandAction;
}

export interface DirectoryTreeState {
  expandedKeys?: Key[];
  selectedKeys?: Key[];
}

function getIcon(props: OcTreeNodeAttribute): React.ReactNode {
  const { isLeaf, expanded } = props;
  if (isLeaf) {
    return (
      <Icon path={IconName.mdiFileDocumentOutline} size={IconSize.Small} />
    );
  }
  return expanded ? (
    <Icon path={IconName.mdiFolderOpenOutline} size={IconSize.Small} />
  ) : (
    <Icon path={IconName.mdiFolderOutline} size={IconSize.Small} />
  );
}

function getTreeData({ treeData, children }: DirectoryTreeProps) {
  return treeData || convertTreeToData(children);
}

const DirectoryTree: React.ForwardRefRenderFunction<
  OcTree,
  DirectoryTreeProps
> = (
  { defaultExpandAll, defaultExpandParent, defaultExpandedKeys, ...props },
  ref
) => {
  // Shift click usage
  const lastSelectedKey = React.useRef<Key>();

  const cachedSelectedKeys = React.useRef<Key[]>();

  const treeRef = React.createRef<OcTree>();

  React.useImperativeHandle(ref, () => treeRef.current!);

  const getInitExpandedKeys = () => {
    const { keyEntities } = convertDataToEntities(getTreeData(props));

    let initExpandedKeys: any;

    // Expanded keys
    if (defaultExpandAll) {
      initExpandedKeys = Object.keys(keyEntities);
    } else if (defaultExpandParent) {
      initExpandedKeys = conductExpandParent(
        props.expandedKeys || defaultExpandedKeys || [],
        keyEntities
      );
    } else {
      initExpandedKeys = props.expandedKeys || defaultExpandedKeys;
    }
    return initExpandedKeys;
  };

  const [selectedKeys, setSelectedKeys] = React.useState(
    props.selectedKeys || props.defaultSelectedKeys || []
  );
  const [expandedKeys, setExpandedKeys] = React.useState(getInitExpandedKeys());

  React.useEffect(() => {
    if ('selectedKeys' in props) {
      setSelectedKeys(props.selectedKeys!);
    }
  }, [props.selectedKeys]);

  React.useEffect(() => {
    if ('expandedKeys' in props) {
      setExpandedKeys(props.expandedKeys);
    }
  }, [props.expandedKeys]);

  const expandFolderNode = (
    event: React.MouseEvent<HTMLElement>,
    node: any
  ) => {
    const { isLeaf } = node;

    if (isLeaf || event.shiftKey || event.metaKey || event.ctrlKey) {
      return;
    }

    // Call internal tree expand function
    debounce(() => {
      treeRef.current!.onNodeExpand(event as any, node);
    }, 200);
  };

  const onExpand = (
    keys: Key[],
    info: {
      node: EventDataNode<DataNode>;
      expanded: boolean;
      nativeEvent: MouseEvent;
    }
  ) => {
    if (!('expandedKeys' in props)) {
      setExpandedKeys(keys);
    }
    // Call origin function
    return props.onExpand?.(keys, info);
  };

  const onClick = (
    event: React.MouseEvent<HTMLElement>,
    node: EventDataNode<DataNode>
  ) => {
    const { expandAction } = props;

    // Expand the tree
    if (expandAction === 'click') {
      expandFolderNode(event, node);
    }

    props.onClick?.(event, node);
  };

  const onDoubleClick = (
    event: React.MouseEvent<HTMLElement>,
    node: EventDataNode<DataNode>
  ) => {
    const { expandAction } = props;

    // Expand the tree
    if (expandAction === 'doubleClick') {
      expandFolderNode(event, node);
    }

    props.onDoubleClick?.(event, node);
  };

  const onSelect = (
    keys: Key[],
    event: {
      event: 'select';
      selected: boolean;
      node: any;
      selectedNodes: DataNode[];
      nativeEvent: MouseEvent;
    }
  ) => {
    const { multiple } = props;
    const { node, nativeEvent } = event;
    const { key = '' } = node;

    const treeData = getTreeData(props);
    // const newState: DirectoryTreeState = {};

    // We need wrap this event since some value is not same
    const newEvent: any = {
      ...event,
      selected: true, // Directory selected always true
    };

    // Windows / Mac single pick
    const ctrlPick: boolean = nativeEvent?.ctrlKey || nativeEvent?.metaKey;
    const shiftPick: boolean = nativeEvent?.shiftKey;

    // Generate new selected keys
    let newSelectedKeys: Key[];
    if (multiple && ctrlPick) {
      // Control click
      newSelectedKeys = keys;
      lastSelectedKey.current = key;
      cachedSelectedKeys.current = newSelectedKeys;
      newEvent.selectedNodes = convertDirectoryKeysToNodes(
        treeData,
        newSelectedKeys
      );
    } else if (multiple && shiftPick) {
      // Shift click
      newSelectedKeys = Array.from(
        new Set([
          ...(cachedSelectedKeys.current || []),
          ...calcRangeKeys({
            treeData,
            expandedKeys,
            startKey: key,
            endKey: lastSelectedKey.current,
          }),
        ])
      );
      newEvent.selectedNodes = convertDirectoryKeysToNodes(
        treeData,
        newSelectedKeys
      );
    } else {
      // Single click
      newSelectedKeys = [key];
      lastSelectedKey.current = key;
      cachedSelectedKeys.current = newSelectedKeys;
      newEvent.selectedNodes = convertDirectoryKeysToNodes(
        treeData,
        newSelectedKeys
      );
    }

    props.onSelect?.(newSelectedKeys, newEvent);
    if (!('selectedKeys' in props)) {
      setSelectedKeys(newSelectedKeys);
    }
  };
  const htmlDir: string = useCanvasDirection();

  const { classNames, ...otherProps } = props;

  const connectClassNames = mergeClasses([
    styles.treeDirectory,
    { ['tree-directory-rtl']: htmlDir === 'rtl' },
    classNames,
  ]);

  return (
    <Tree
      icon={getIcon}
      ref={treeRef}
      blockNode
      {...otherProps}
      classNames={connectClassNames}
      expandedKeys={expandedKeys}
      selectedKeys={selectedKeys}
      onSelect={onSelect}
      onClick={onClick}
      onDoubleClick={onDoubleClick}
      onExpand={onExpand}
    />
  );
};

const ForwardDirectoryTree = React.forwardRef(DirectoryTree);
ForwardDirectoryTree.displayName = 'DirectoryTree';

ForwardDirectoryTree.defaultProps = {
  showIcon: true,
  expandAction: 'click' as DirectoryTreeProps['expandAction'],
};

export default ForwardDirectoryTree;
