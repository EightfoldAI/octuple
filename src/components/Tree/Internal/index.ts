import Tree from './OcTree';
import TreeNode from './TreeNode';
import type {
  BasicDataNode,
  FieldDataNode,
  TreeNodeProps,
  OcTreeProps,
} from './OcTree.types';

export { TreeNode };
export type { OcTreeProps, TreeNodeProps, BasicDataNode, FieldDataNode };
export default Tree;
