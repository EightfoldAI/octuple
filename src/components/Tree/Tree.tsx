'use client';

import React from 'react';
import OcTree, { TreeNode } from './Internal';
import { DraggableConfig, OcTreeNodeProps, TreeProps } from './Tree.types';
import type { BasicDataNode, DataNode } from './Internal/OcTree.types';
import DirectoryTree from './DirectoryTree';
import renderSwitcherIcon from './Utils/iconUtil';
import dropIndicatorRender from './Utils/dropIndicator';
import { Icon, IconName, IconSize } from '../Icon';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';
import { collapseMotion, mergeClasses } from '../../shared/utilities';

import styles from './Styles/tree.module.scss';

type CompoundedComponent = (<T extends BasicDataNode | DataNode = DataNode>(
  props: React.PropsWithChildren<TreeProps<T>> & { ref?: React.Ref<OcTree> }
) => React.ReactElement) & {
  defaultProps: Partial<React.PropsWithChildren<TreeProps<any>>>;
  TreeNode: typeof TreeNode;
  DirectoryTree: typeof DirectoryTree;
};

const Tree = React.forwardRef<OcTree, TreeProps>((props, ref) => {
  const {
    classNames,
    showIcon,
    showLine,
    switcherIcon,
    blockNode,
    children,
    checkable,
    selectable,
    draggable,
    virtual,
  } = props;
  const htmlDir: string = useCanvasDirection();
  const newProps = {
    ...props,
    showLine: Boolean(showLine),
    dropIndicatorRender,
  };

  const draggableConfig = React.useMemo(() => {
    if (!draggable) {
      return false;
    }

    let mergedDraggable: DraggableConfig = {};
    switch (typeof draggable) {
      case 'function':
        mergedDraggable.nodeDraggable = draggable;
        break;

      case 'object':
        mergedDraggable = { ...draggable };
        break;

      default:
      // Do nothing
    }

    if (mergedDraggable.icon !== false) {
      mergedDraggable.icon = mergedDraggable.icon || (
        <Icon
          color={'#ffffff'}
          path={IconName.mdiHandBackRight}
          size={IconSize.Medium}
        />
      );
    }

    return mergedDraggable;
  }, [draggable]);

  return (
    <OcTree
      itemHeight={20}
      ref={ref}
      virtual={virtual}
      {...newProps}
      classNames={mergeClasses([
        { ['tree-icon-hide']: !showIcon },
        { [styles.treeBlockNode]: blockNode },
        { [styles.treeUnselectable]: !selectable },
        { [styles.treeRtl]: htmlDir === 'rtl' },
        classNames,
      ])}
      direction={htmlDir}
      checkable={
        checkable ? <span className={'tree-checkbox-inner'} /> : checkable
      }
      selectable={selectable}
      switcherIcon={(nodeProps: OcTreeNodeProps) =>
        renderSwitcherIcon(switcherIcon, showLine, nodeProps)
      }
      draggable={draggableConfig as any}
    >
      {children}
    </OcTree>
  );
}) as unknown as CompoundedComponent;

Tree.TreeNode = TreeNode;

Tree.DirectoryTree = DirectoryTree;

Tree.defaultProps = {
  checkable: false,
  selectable: true,
  showIcon: false,
  motion: {
    ...collapseMotion,
    motionAppear: false,
  },
  blockNode: false,
};

export default Tree;
