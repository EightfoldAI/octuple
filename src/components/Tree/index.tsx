import Tree from './Tree';

export { EventDataNode, DataNode } from './Internal/OcTree.types';

export {
    TreeProps,
    OcTreeNode,
    OcTreeNodeMouseEvent,
    OcTreeNodeExpandedEvent,
    OcTreeNodeCheckedEvent,
    OcTreeNodeSelectedEvent,
    OcTreeNodeAttribute,
    OcTreeNodeProps,
} from './Tree.types';

export {
    ExpandAction as DirectoryTreeExpandAction,
    DirectoryTreeProps,
} from './DirectoryTree';

export default Tree;
