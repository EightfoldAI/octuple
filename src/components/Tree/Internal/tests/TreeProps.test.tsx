import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Tree, { TreeNode, FieldDataNode } from '..';
import { spyConsole, spyError } from './util';

// Promisify timeout to let jest catch works
function timeoutPromise(delay = 0) {
    return new Promise((resolve) => {
        setTimeout(resolve, delay);
    });
}

describe('Tree Props', () => {
    spyConsole();

    // showLine
    it('showLine', () => {
        const wrapper = render(<Tree showLine />);
        expect(wrapper.container.firstChild).toMatchSnapshot();
    });

    // showIcon
    it('showIcon', () => {
        const withIcon = render(
            <Tree>
                <TreeNode>
                    <TreeNode>
                        <TreeNode />
                    </TreeNode>
                    <TreeNode />
                </TreeNode>
                <TreeNode />
            </Tree>
        );
        expect(withIcon.container.firstChild).toMatchSnapshot();

        const withoutIcon = render(
            <Tree showIcon={false}>
                <TreeNode>
                    <TreeNode>
                        <TreeNode />
                    </TreeNode>
                    <TreeNode />
                </TreeNode>
                <TreeNode />
            </Tree>
        );
        expect(withoutIcon.container.firstChild).toMatchSnapshot();

        const withOpenIcon = render(
            <Tree defaultExpandedKeys={['0-0']}>
                <TreeNode>
                    <TreeNode key="0-0">
                        <TreeNode />
                    </TreeNode>
                    <TreeNode />
                </TreeNode>
                <TreeNode />
            </Tree>
        );
        expect(withOpenIcon.container.firstChild).toMatchSnapshot();
    });

    describe('selectable', () => {
        // selectable - false
        it('without selectable', () => {
            const handleOnSelect = jest.fn();
            const handleOnCheck = jest.fn();

            const { container } = render(
                <Tree
                    onSelect={handleOnSelect}
                    defaultExpandedKeys={['0-0']}
                    selectable={false}
                >
                    <TreeNode key="0-0">
                        <TreeNode key="0-0-0" />
                    </TreeNode>
                </Tree>
            );

            expect(container.firstChild).toMatchSnapshot();

            fireEvent.click(container.querySelector('.treeNodeContentWrapper'));

            expect(handleOnSelect).not.toHaveBeenCalled();
            expect(handleOnCheck).not.toHaveBeenCalled(); // Will test in checkable
        });
    });

    describe('expandAction with selectable props', () => {
        it('title expandable when selectable is false and expandAction is "click"', () => {
            const onClick = jest.fn();
            const onSelect = jest.fn();
            const onExpand = jest.fn();

            const { container } = render(
                <Tree
                    onClick={onClick}
                    onSelect={onSelect}
                    onExpand={onExpand}
                    defaultExpandedKeys={['0-0']}
                    selectable={false}
                    expandAction="click"
                >
                    <TreeNode title="parent 1" key="0-0">
                        <TreeNode title="leaf 1" key="0-0-0">
                            <TreeNode title="leaf-1" key="0-0-0-0" />
                        </TreeNode>

                        <TreeNode title="leaf 2" key="0-1-0">
                            <TreeNode title="leaf-2" key="0-1-0-0" />
                        </TreeNode>
                    </TreeNode>
                </Tree>
            );

            // test trigger expand when click title
            fireEvent.click(container.querySelector('[title="leaf 1"]'));

            expect(onClick).toHaveBeenCalled();
            expect(onSelect).not.toHaveBeenCalled();
            expect(onExpand).toHaveBeenCalledWith(['0-0', '0-0-0'], {
                expanded: true,
                node: expect.anything(),
                nativeEvent: expect.anything(),
            });

            onClick.mockReset();
            onSelect.mockReset();
            onExpand.mockReset();

            // test trigger un-expand when click title again
            fireEvent.click(container.querySelector('[title="leaf 1"]'));

            expect(onClick).toHaveBeenCalled();
            expect(onSelect).not.toHaveBeenCalled();
            expect(onExpand).toHaveBeenCalledWith(['0-0'], {
                expanded: false,
                node: expect.anything(),
                nativeEvent: expect.anything(),
            });
        });

        it('title expandable when selectable is false and expandAction is "doubleClick"', () => {
            const onDoubleClick = jest.fn();
            const onSelect = jest.fn();
            const onExpand = jest.fn();

            const { container } = render(
                <Tree
                    onDoubleClick={onDoubleClick}
                    onSelect={onSelect}
                    onExpand={onExpand}
                    defaultExpandedKeys={['0-0']}
                    selectable={false}
                    expandAction="doubleClick"
                >
                    <TreeNode title="parent 1" key="0-0">
                        <TreeNode title="leaf 1" key="0-0-0">
                            <TreeNode title="leaf-1" key="0-0-0-0" />
                        </TreeNode>

                        <TreeNode title="leaf 2" key="0-1-0">
                            <TreeNode title="leaf-2" key="0-1-0-0" />
                        </TreeNode>
                    </TreeNode>
                </Tree>
            );

            // test trigger expand when double click title
            fireEvent.doubleClick(container.querySelector('[title="leaf 1"]'));

            expect(onDoubleClick).toHaveBeenCalled();
            expect(onSelect).not.toHaveBeenCalled();
            expect(onExpand).toHaveBeenCalledWith(['0-0', '0-0-0'], {
                expanded: true,
                node: expect.anything(),
                nativeEvent: expect.anything(),
            });

            onDoubleClick.mockReset();
            onSelect.mockReset();
            onExpand.mockReset();

            // test trigger un-expand when double click title again
            fireEvent.doubleClick(container.querySelector('[title="leaf 1"]'));

            expect(onDoubleClick).toHaveBeenCalled();
            expect(onSelect).not.toHaveBeenCalled();
            expect(onExpand).toHaveBeenCalledWith(['0-0'], {
                expanded: false,
                node: expect.anything(),
                nativeEvent: expect.anything(),
            });
        });

        it('title un-expandable when selectable is false and expandAction is false', () => {
            const onClick = jest.fn();
            const onSelect = jest.fn();
            const onExpand = jest.fn();

            const { container } = render(
                <Tree
                    onClick={onClick}
                    onSelect={onSelect}
                    onExpand={onExpand}
                    defaultExpandedKeys={['0-0']}
                    selectable={false}
                    expandAction={false}
                >
                    <TreeNode title="parent 1" key="0-0">
                        <TreeNode title="leaf 1" key="0-0-0">
                            <TreeNode title="leaf-1" key="0-0-0-0" />
                        </TreeNode>

                        <TreeNode title="leaf 2" key="0-1-0">
                            <TreeNode title="leaf-2" key="0-1-0-0" />
                        </TreeNode>
                    </TreeNode>
                </Tree>
            );

            // test won't trigger expand when click title if expandAction is false
            fireEvent.click(container.querySelector('[title="leaf 2"]'));

            expect(onClick).toHaveBeenCalled();
            expect(onSelect).not.toHaveBeenCalled();
            expect(onExpand).not.toHaveBeenCalled();
        });
    });

    // checkable
    describe('checkable', () => {
        it('node set checkable to `false`', () => {
            const { container } = render(
                <Tree checkable defaultExpandAll>
                    <TreeNode key="0-0">
                        <TreeNode key="0-0-0" checkable={false} />
                    </TreeNode>
                </Tree>
            );

            expect(container.querySelectorAll('.treeCheckbox')).toHaveLength(1);
        });
    });

    // Don't crash
    describe('invalidate checkedKeys', () => {
        const errorSpy = spyError();

        const genWrapper = (checkedKeys: any) =>
            render(
                <Tree checkedKeys={checkedKeys} defaultExpandAll checkable>
                    <TreeNode key="0-0">
                        <TreeNode key="0-0-0" />
                    </TreeNode>
                </Tree>
            );

        it('null', () => {
            const { container } = genWrapper(null);
            expect(errorSpy()).not.toHaveBeenCalledWith(
                'Warning: `checkedKeys` is not an array or an object'
            );
            expect(container.firstChild).toMatchSnapshot();
        });
    });

    // defaultExpandAll
    it('defaultExpandAll', () => {
        const { container } = render(
            <Tree defaultExpandAll>
                <TreeNode key="0-0">
                    <TreeNode key="0-0-0" />
                </TreeNode>
            </Tree>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('icon', () => {
        // Node icon has much higher priority
        const { container } = render(
            <Tree defaultExpandAll icon={<span>ROOT ICON</span>}>
                <TreeNode key="0-0">
                    <TreeNode key="0-0-0" icon={<span>CUSTOMIZE ICON</span>} />
                </TreeNode>
            </Tree>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('onClick', () => {
        const onClick = jest.fn();

        const { container } = render(
            <Tree onClick={onClick} defaultExpandedKeys={['0-0']}>
                <TreeNode key="0-0">
                    <TreeNode key="0-0-0" />
                </TreeNode>
            </Tree>
        );

        // Select leaf
        fireEvent.click(
            container.querySelectorAll('.treeNodeContentWrapper')[1]
        );

        expect(onClick).toHaveBeenCalledWith(
            expect.objectContaining({}),
            expect.objectContaining({
                key: '0-0-0',
            })
        );
    });

    it('onDoubleClick', () => {
        const onClick = jest.fn();
        const onDoubleClick = jest.fn();

        const { container } = render(
            <Tree
                onClick={onClick}
                onDoubleClick={onDoubleClick}
                defaultExpandedKeys={['0-0']}
            >
                <TreeNode key="0-0">
                    <TreeNode key="0-0-0" />
                </TreeNode>
            </Tree>
        );

        // Select leaf
        fireEvent.doubleClick(
            container.querySelectorAll('.treeNodeContentWrapper')[1]
        );

        expect(onClick).not.toHaveBeenCalled();
        expect(onDoubleClick).toHaveBeenCalledWith(
            expect.objectContaining({}),
            expect.objectContaining({
                key: '0-0-0',
            })
        );
    });

    it('onContextMenu', () => {
        const onClick = jest.fn();
        const onContextMenu = jest.fn();

        const { container } = render(
            <Tree
                onClick={onClick}
                onContextMenu={onContextMenu}
                defaultExpandedKeys={['0-0']}
            >
                <TreeNode key="0-0">
                    <TreeNode key="0-0-0" />
                </TreeNode>
            </Tree>
        );

        // Select leaf
        fireEvent.contextMenu(
            container.querySelectorAll('.treeNodeContentWrapper')[1]
        );

        expect(onClick).not.toHaveBeenCalled();
        expect(onContextMenu).toHaveBeenCalled();
    });

    describe('loadedKeys & onLoad', () => {
        it('has loadedKeys', () => {
            const loadData = jest.fn(() => Promise.resolve());
            const onLoad = jest.fn();

            const { container } = render(
                <Tree loadedKeys={['0-0']} loadData={loadData} onLoad={onLoad}>
                    <TreeNode key="0-0" />
                </Tree>
            );

            fireEvent.click(container.querySelector('.treeSwitcher'));
            expect(loadData).not.toHaveBeenCalled();
            expect(onLoad).not.toHaveBeenCalled();
        });
    });

    it('treeData', () => {
        const treeData = [
            { key: 'K0', title: 'T0' },
            {
                key: 'K1',
                title: 'T1',
                children: [
                    { key: 'K10', title: 'T10' },
                    {
                        key: 'K11',
                        title: 'T11',
                        children: [
                            { key: 'K110', title: 'T110' },
                            { key: 'K111', title: 'T111' },
                        ],
                    },
                    { key: 'K12', title: 'T12' },
                ],
            },
        ];
        const { container, rerender } = render(
            <Tree treeData={treeData} defaultExpandAll />
        );
        expect(container.firstChild).toMatchSnapshot();

        rerender(
            <Tree treeData={[{ key: 'K0', title: 'T0' }]} defaultExpandAll />
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    describe('disabled', () => {
        it('basic', () => {
            const { container } = render(
                <Tree defaultExpandAll disabled>
                    <TreeNode key="0-0" />
                </Tree>
            );
            expect(container.firstChild).toMatchSnapshot();
        });

        it('treeNode should disabled when tree disabled', () => {
            const { container } = render(
                <Tree defaultExpandAll disabled>
                    <TreeNode key="0-0" disabled={false} />
                </Tree>
            );
            expect(container.firstChild).toMatchSnapshot();
        });
    });

    describe('data and aria props', () => {
        it('renders data attributes', () => {
            const { container } = render(<Tree data-test="tree" />);
            expect(container.firstChild).toMatchSnapshot();
        });

        it('renders aria attributes', () => {
            const { container } = render(<Tree aria-label="name" />);
            expect(container.firstChild).toMatchSnapshot();
        });
    });

    describe('custom switcher icon', () => {
        function switcherIcon(text: React.ReactNode, testLeaf?: boolean) {
            const sfc = (props: { isLeaf: boolean }) => {
                if (testLeaf) {
                    return props.isLeaf ? <span>{text}</span> : null;
                }
                return props.isLeaf ? null : <span>{text}</span>;
            };

            return sfc;
        }

        it('switcher icon', () => {
            const { container } = render(
                <Tree
                    defaultExpandAll
                    switcherIcon={switcherIcon('switcherIcon')}
                >
                    <TreeNode key="0-0" />
                    <TreeNode
                        key="0-1"
                        switcherIcon={switcherIcon('switcherIconFromNode0-1')}
                    >
                        <TreeNode key="0-1-0" />
                        <TreeNode key="0-1-1" />
                    </TreeNode>
                </Tree>
            );
            expect(container.firstChild).toMatchSnapshot();
        });

        it('switcher leaf icon', () => {
            const { container } = render(
                <Tree
                    defaultExpandAll
                    switcherIcon={switcherIcon('switcherLeafIcon', true)}
                >
                    <TreeNode key="0-0" />
                    <TreeNode
                        key="0-1"
                        switcherIcon={switcherIcon(
                            'switcherLeafIconFromNode0-1',
                            true
                        )}
                    />
                    <TreeNode key="0-2">
                        <TreeNode key="0-2-0" />
                        <TreeNode
                            key="0-2-1"
                            switcherIcon={switcherIcon(
                                'switcherLeafIconFromNode0-2-1',
                                true
                            )}
                        />
                    </TreeNode>
                    <TreeNode key="0-3" />
                </Tree>
            );
            expect(container.firstChild).toMatchSnapshot();
        });
    });

    it('titleRender', () => {
        const { container } = render(
            <Tree<FieldDataNode<{ value: string; title?: any }>>
                defaultExpandAll
                titleRender={({ value }) => (
                    <span className="bamboo-span">{value}</span>
                )}
                treeData={[
                    {
                        title: (props: { value: any }) => (
                            <span className="light-span">{props.value}</span>
                        ),
                        value: 'light',
                    },
                    { value: 'bamboo' },
                ]}
            />
        );

        expect(
            container
                .querySelectorAll('.treeTitle')[0]
                .querySelector('.light-span').textContent
        ).toEqual('light');
        expect(
            container
                .querySelectorAll('.treeTitle')[1]
                .querySelector('.bamboo-span').textContent
        ).toEqual('bamboo');
    });
});
