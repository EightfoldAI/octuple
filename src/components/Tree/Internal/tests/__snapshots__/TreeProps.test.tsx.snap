version https://git-lfs.github.com/spec/v1
oid sha256:324075355fa4973786b258d105ac101a458cfbc96cf7831c1e2d396352aa2770
size 44090
